openedx/features/content_type_gating/partitions.py

```diff
@@ -0,0 +1,192 @@
+    the gated content despite not being verified users.
+    """
+
+    LIMITED_ACCESS = Group(settings.CONTENT_TYPE_GATE_GROUP_IDS['limited_access'], 'Limited-access Users')
+    FULL_ACCESS = Group(settings.CONTENT_TYPE_GATE_GROUP_IDS['full_access'], 'Full-access Users')
+
+    @classmethod
+    def get_group_for_user(cls, course_key, user, user_partition, **kwargs):
+        """
+        Returns the Group for the specified user.
+        """
+
+        # First, check if we have to deal with masquerading.
+        # If the current user is masquerading as a specific student, use the
+        # same logic as normal to return that student's group. If the current
+        # user is masquerading as a generic student in a specific group, then
+        # return that group.
+        if get_course_masquerade(user, course_key) and not is_masquerading_as_specific_student(user, course_key):
+            return get_masquerading_user_group(course_key, user, user_partition)
+
+        # For now, treat everyone as a Full-access user, until we have the rest of the
+        # feature gating logic in place.
+
+        if not CONTENT_TYPE_GATING_FLAG.is_enabled():
+            return cls.FULL_ACCESS
+
+        # If CONTENT_TYPE_GATING is enabled use the following logic to determine whether a user should have FULL_ACCESS
+        # or LIMITED_ACCESS
+
+        course_mode = apps.get_model('course_modes.CourseMode')
+        modes = course_mode.modes_for_course_dict(course_key)
+
+        # If there is no verified mode, all users are granted FULL_ACCESS
+        if not course_mode.has_verified_mode(modes):
+            return cls.FULL_ACCESS
+
+        course_enrollment = apps.get_model('student.CourseEnrollment')
+
+        # TODO: remove this line
+        # enrollment = course_enrollment.get_enrollment(user, course_key)
+
+        mode_slug, is_active = course_enrollment.enrollment_mode_for_user(user, course_key)
+
+        if mode_slug and is_active:
+            course_mode = course_mode.mode_for_course(
+                course_key,
+                mode_slug,
+                modes=course_mode.modes_for_course(course_key, include_expired=True, only_selectable=False),
+            )
+            if course_mode is None:
+                LOG.error(
+                    "User %s is in an unknown CourseMode '%s' for course %s. Granting full access to content for this user",
+                    user.username,
+                    mode_slug,
+                    course_key,
+                )
+                return cls.FULL_ACCESS
+
+            if mode_slug == COURSE_MODE_AUDIT:
+                # TODO: does the below comment mean there is more work to be done here
+                # Check the user email exceptions here:
+                return cls.LIMITED_ACCESS
+            else:
+                return cls.FULL_ACCESS
+        else:
+            # Unenrolled users don't get gated content
```

Comment:
what is the scenario that would hit this code path?

Reply:
I was more just curious if this would be unreachable code since I'm assuming that unenrolled users can't see course content.
However, it does probably make sense to still handle this scenario and put unenrolled users in the limited access group.  
For example, there is long-running project to make courseware content indexable by google in which case we would want the gated problems to be gated.

Reply:
I think theoretically an unenrolled user could hit the URL of a piece of gated content, but a user shouldn't hit this scenario under normal circumstances.